Update named colors list to match CSS Color 4

weinig opened the pull request:
#### 09df074a4a8861c5fa7ba160368f6fb1ea1890b0
<pre>
Update named colors list to match CSS Color 4
<a href="https://bugs.webkit.org/show_bug.cgi?id=245533">https://bugs.webkit.org/show_bug.cgi?id=245533</a>
&lt;rdar://100287586&gt;

Reviewed by Darin Adler.

Added new system colors from CSS Color 4 with the following default
platform agnostic values:

  ButtonBorder
    white (matching existing ActiveBorder and InactiveBorder values).
  SelectedItem
    lightGray (matching existing Menu value)
  SelectedItemText
    black (matching existing *Text values)
  Mark
    yellow (matching background-color of &lt;mark&gt; in UA stylesheet)
  MarkText
    black (matching color of &lt;mark&gt; in UA stylesheet)
  AccentColor
    sRGB { 0, 122, 255 } (matching color of system-control-accent)
  AccentColorText
    black (matching existing *Text values)

As with other system colors, it may make sense to specialize
these per-platform, but this change does not aim to do that.

Also re-organized CSSValueKeywords.in to group some related
color names together, and added comments to help clarify what
set things belong to.

As a way of organizing, and in preparation for an upcoming
change to deprecated system colors, the deprecated system
colors were grouped together and a helper function added.

As we now have a named colors Mark and MarkText, the UA stylesheet
for &lt;mark&gt; was updated to use the named colors (no observable
change in behavior though, as the colors still resolve to yellow
and black).

* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/StyleColor.cpp:
(WebCore::isVGAPaletteColor):
(WebCore::StyleColor::isAbsoluteColorKeyword):
(WebCore::StyleColor::isSystemColorKeyword):
(WebCore::StyleColor::isDeprecatedSystemColorKeyword):
* Source/WebCore/css/StyleColor.h:
* Source/WebCore/css/html.css:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::systemColor const):

Canonical link: <a href="https://commits.webkit.org/254840@main">https://commits.webkit.org/254840@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5cffb7fe5fe41e3e0ac8c3d5a965c8b3242eef3a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90414 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35001 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/21025 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99748 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/157213 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33499 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28706 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82777 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/96193 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96069 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26628 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77264 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26503 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81426 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/81212 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69515 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/34593 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15263 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32417 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16226 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36180 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/39174 "Found 1 new test failure: css3/scroll-snap/scroll-snap-drag-scrollbar-thumb-with-relayouts.html (failure)") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1462 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38094 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35315 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1321,83 +1321,246 @@ auto RenderTheme::colorCache(OptionSet<StyleColorOptions> options) const -> Colo
+    case CSSValueButtonface:
+        return Color::lightGray;
+
+    // https://drafts.csswg.org/css-color-4/#valdef-system-color-buttontext
+    // Text on push buttons.
+    case CSSValueButtontext:
+        return Color::black;
+
+    // https://drafts.csswg.org/css-color-4/#valdef-system-color-buttonborder
+    // The base border color for push buttons.
+    case CSSValueButtonborder:
+        return Color::white;
```
These reflect the view that "you can make a standard button using CSS", which obviously seems untrue for the style of buttons on Apple platforms. I suppose then the color values are arbitrary since these can’t be used to make buttons that conform to the platform standards for what buttons look like?

weinig replied:
I do really question why these exist (and are being added to) these days, as this concept really harkens back to a very Windows centric web (most of the existing names are straight out of the Win32 SDK). I really don't know what these could be used for in practice, as there just aren't enough of them to fully express what platform buttons look like (on macOS I am pretty sure `<button>` has shadow/gradient that defines the border for instance). 

There are some WPT tests that seem to suggest there is some expectation of these values matching UA stylesheet values, but I can't find any specification for that, and WebKit has really never had this property so I doubt it is required for compat.  

I filed https://github.com/w3c/csswg-drafts/issues/7781 to ask the editors about it.

On the other hand, we already support so many of these, that adding these final few hardly seems like it will tip the scales in any direction.

darinadler commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1321,83 +1321,246 @@ auto RenderTheme::colorCache(OptionSet<StyleColorOptions> options) const -> Colo
+    case CSSValueFieldtext:
+        return Color::black;
+
+    // https://drafts.csswg.org/css-color-4/#valdef-system-color-highlight
+    // Background of selected text, for example from ::selection.
+    case CSSValueHighlight:
+        return SRGBA<uint8_t> { 181, 213, 255 };
```
I am a little bit confused; the old code does this same thing, so this is not a change. But macOS has a highlight color that is chosen in the System Settings app. Should this be exposed here? Seems like a fingerprinting risk, but not exposing it seems like a correctness risk.

xeenon replied:
Indeed it would be a fingerprinting risk to expose the system color. To mitigate that and still allow native UI to use the accent derived colors we tied that to the WKWebView `_useSystemAppearance` SPI. When that is `NO` we return fixed blue derived colors only. When it is `YES` we use the `NSColor` / `UIColor` APIs to return real accent derived colors.

That conditional treatment should be done here for these new colors.

weinig replied:
> I am a little bit confused; the old code does this same thing, so this is not a change. But macOS has a highlight color that is chosen in the System Settings app. Should this be exposed here? Seems like a fingerprinting risk, but not exposing it seems like a correctness risk.

This is existing code, I just reformatted and re-ordered it.

As Tim notes, we actually do have an override of this in RenderThemeMac/RenderThemeIOS that does a more fingerprinty thing when a native might want it.

darinadler commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1321,83 +1321,246 @@ auto RenderTheme::colorCache(OptionSet<StyleColorOptions> options) const -> Colo
+    case CSSValueSelecteditemtext:
+        return Color::black;
+
+    // https://drafts.csswg.org/css-color-4/#valdef-system-color-mark
+    // Background of text that has been specially marked (such as by the HTML mark element).
+    case CSSValueMark:
+        return Color::yellow;
```
Given the existence of `RenderTheme::platformAnnotationHighlightColor` I am not sure what this should be returning.

pxlcoder replied:
I think we should return `platformAnnotationHighlightColor` here.

weinig replied:
That would be a change in behavior for `<mark>`, so I'd prefer to leave that to a subsequent change.

darinadler commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1321,83 +1321,246 @@ auto RenderTheme::colorCache(OptionSet<StyleColorOptions> options) const -> Colo
+    case CSSValueGraytext:
+        return Color::darkGray;
+
+    // https://drafts.csswg.org/css-color-4/#valdef-system-color-accentcolor
+    // Background of accented user interface controls.
+    case CSSValueAccentcolor:
+        return SRGBA<uint8_t> { 0, 122, 255 };
```
No idea what an "accented" user interface control is. Perhaps a concept that does not exist on Apple platforms?

pxlcoder replied:
This would be the system accent color on Apple platforms, which is configurable on macOS using `Appearance > Accent Color` in System Settings.

Something we would want to force to a single value even when implementing the platform-specific version, to avoid fingerprinting risk.

weinig replied:
Yeah, we have an existing webkit specific named color, CSSValueAppleSystemControlAccent, which uses this same hard coded value when not in the special "allow system UI" mode Tim mentioned above.

In a follow up, we should consider making CSSValueAccentcolor have the same behavior as CSSValueAppleSystemControlAccent.

darinadler commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1321,83 +1321,246 @@ auto RenderTheme::colorCache(OptionSet<StyleColorOptions> options) const -> Colo
+    case CSSValueMarktext:
+        return Color::black;
+
+    // https://drafts.csswg.org/css-color-4/#valdef-system-color-graytext
+    // Disabled text. (Often, but not necessarily, gray.)
+    case CSSValueGraytext:
+        return Color::darkGray;
```
Does this need to match the color of disabled text in system controls?

pxlcoder replied:
I think using `RenderTheme::disabledTextColor` with black/white would be better here. It would match the disabled text color used in our form controls.

weinig replied:
This is existing, just shifted. I don't see a great reason to change it in this PR. 

As a side note, on macOS, CSSValueGraytext uses [NSColor disabledControlTextColor:] already.

darinadler reviewed (approved):
Did my best to review. It’s not clear to me whether our goal for named colors should be to fit in really well with the surrounding operating system, or rather to just choose values same as other browsers to decrease the chances of any interoperability problems and to keep fingerprinting risk to a minimum.

Commit:
Update named colors list to match CSS Color 4
https://bugs.webkit.org/show_bug.cgi?id=245533
<rdar://100287586>

Reviewed by Darin Adler.

Added new system colors from CSS Color 4 with the following default
platform agnostic values:

  ButtonBorder
    white (matching existing ActiveBorder and InactiveBorder values).
  SelectedItem
    lightGray (matching existing Menu value)
  SelectedItemText
    black (matching existing *Text values)
  Mark
    yellow (matching background-color of <mark> in UA stylesheet)
  MarkText
    black (matching color of <mark> in UA stylesheet)
  AccentColor
    sRGB { 0, 122, 255 } (matching color of system-control-accent)
  AccentColorText
    black (matching existing *Text values)

As with other system colors, it may make sense to specialize
these per-platform, but this change does not aim to do that.

Also re-organized CSSValueKeywords.in to group some related
color names together, and added comments to help clarify what
set things belong to.

As a way of organizing, and in preparation for an upcoming
change to deprecated system colors, the deprecated system
colors were grouped together and a helper function added.

As we now have a named colors Mark and MarkText, the UA stylesheet
for <mark> was updated to use the named colors (no observable
change in behavior though, as the colors still resolve to yellow
and black).

* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/StyleColor.cpp:
(WebCore::isVGAPaletteColor):
(WebCore::StyleColor::isAbsoluteColorKeyword):
(WebCore::StyleColor::isSystemColorKeyword):
(WebCore::StyleColor::isDeprecatedSystemColorKeyword):
* Source/WebCore/css/StyleColor.h:
* Source/WebCore/css/html.css:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::systemColor const):

Canonical link: https://commits.webkit.org/254840@main
